### DISCUSSION TITLE
How to build a feature with drivers like the storage

### DISCUSSION THREAD

**Discussion by Barbapapazes** [nitrojs/nitro#2737] [Help Wanted]
Hello,

I'm building a feature with drivers. I want to do something similar to the unstorage integration but because nothing from the rollup configuration is exposed, it's very difficult to add a virtual module through the rollup plugin system.

How can I do this?

**Comment by Barbapapazes** [comment]
Maybe I should try to use a nightly version to have access to the rollup parts.

**Reply by Barbapapazes** [reply]
Nop, only access to `getRollupConfig`

**Comment by TheAlexLichter** [comment]
What stops you from using the [rollupConfig option directly](https://github.com/unjs/nitro/blob/e28a02f7675e21b3aaa100d4f41f50ef1e4cb6a7/src/rollup/config.ts#L129) or the [rollup:before hook](https://github.com/unjs/nitro/blob/e28a02f7675e21b3aaa100d4f41f50ef1e4cb6a7/src/core/build/build.ts#L8)?

**Reply by Barbapapazes** [reply]
Nothing! I will try to do it manually. I was looking to reuse some already existing parts.

**Reply by Barbapapazes** [reply]
The hook `rollup:before` has not been release yet.

**Comment by pi0** [comment]
Have hot tried built-in vfs config? What are the limitations with it if did?

**Reply by Barbapapazes** [reply]
I'm not understanding how to use VFS 😔

**Reply by Barbapapazes** [reply]
Ok, I understand how and when to use it. Is there a way to generate some types? How do you handle types?

**Reply by Barbapapazes** [reply]
Is using a `types.d.ts` at the project root is a good thing?

**Reply by Barbapapazes** [reply]
After looking and diving into the Nitro repo, I do something like this:

```ts
export default defineNitroConfig({
  srcDir: "server",
  virtual: {
    "#internal/amboise": () => {
      return `export const data = 1`;
    },
  },
  typescript: {
    tsConfig: {
      compilerOptions: {
        paths: {
          "#internal/amboise": ["../../types/virtual/amboise.d.ts"],
        },
      },
    },
  },
});
```


works really great and seems easy to adapt in a module.
